core/services/relay/evm/read/bindings.go

```
@@ -0,0 +1,281 @@
+	}
+
+	return binding, values.address, nil
+}
+
+func (b *BindingsRegistry) AddReader(contractName, readName string, rdr Reader) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.contractLookup.addReadNameForContract(contractName, readName)
+
+	cb, cbExists := b.contractBindings[contractName]
+	if !cbExists {
+		cb = newContractBinding(contractName)
+		b.contractBindings[contractName] = cb
+	}
+
+	cb.AddReaderNamed(readName, rdr)
+}
+
+// Bind binds contract addresses to contract bindings and read bindings.
+// Bind also registers the common contract polling filter and eventBindings polling filters.
+func (b *BindingsRegistry) Bind(ctx context.Context, reg Registrar, bindings []commontypes.BoundContract) error {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, binding := range bindings {
+		contract, exists := b.contractBindings[binding.Name]
+		if !exists {
+			return fmt.Errorf("%w: no contract named %s", commontypes.ErrInvalidConfig, binding.Name)
+		}
+
+		b.contractLookup.bindAddressForContract(binding.Name, binding.Address)
+
+		if err := errors.Join(
+			contract.Bind(ctx, reg, common.HexToAddress(binding.Address)),
+			contract.BindReaders(ctx, common.HexToAddress(binding.Address)),
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *BindingsRegistry) BatchGetLatestValues(ctx context.Context, request commontypes.BatchGetLatestValuesRequest) (commontypes.BatchGetLatestValuesResult, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	var batchCall BatchCall
+
+	for binding, contractBatch := range request {
+		cb := b.contractBindings[binding.Name]
+
+		for i := range contractBatch {
+			req := contractBatch[i]
+
+			values, ok := b.contractLookup.getContractForReadName(binding.ReadIdentifier(req.ReadName))
+			if !ok {
+				return nil, fmt.Errorf("%w: no method for read name %s", commontypes.ErrInvalidType, binding.ReadIdentifier(req.ReadName))
+			}
+
+			rdr, exists := cb.GetReaderNamed(values.readName)
+			if !exists {
+				continue
+			}
```

Should this throw an error instead? This is an invalid request